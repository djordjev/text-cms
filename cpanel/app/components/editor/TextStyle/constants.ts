export const TEXT_STYLES = [
  { id: 'heading_1', name: 'Heading 1' },
  { id: 'heading_2', name: 'Heading 2' },
  { id: 'heading_3', name: 'Heading 3' },
  { id: 'heading_4', name: 'Heading 4' },
  { id: 'heading_5', name: 'Heading 5' },
  { id: 'heading_7', name: 'Heading 6' },
  { id: 'paragraph', name: 'Paragraph' }
];
