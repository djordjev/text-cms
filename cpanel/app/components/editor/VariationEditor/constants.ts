import { Descendant } from 'slate';

export const INITIAL_VALUE: Descendant[] = [
  {
    type: 'paragraph',
    children: [{ text: 'A line of text in a paragraph.' }]
  }
];
