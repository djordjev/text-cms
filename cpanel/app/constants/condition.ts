import { ConditionDescriptor } from '~/types/condition';

export const DEFAULT_CONDITION_DESCRIPTOR: ConditionDescriptor = ['', '=', ''];
