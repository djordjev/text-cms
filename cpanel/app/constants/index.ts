export const BUTTON_ACTION = '_action';

export const STRATEGY_USER_PASS = 'user-pass';
