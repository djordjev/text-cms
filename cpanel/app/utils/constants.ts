export const COLORS = {
  BLACK: '#000',
  PRIMARY: '#003EB3'
};

export const BUTTON_ACTION = '_action';
