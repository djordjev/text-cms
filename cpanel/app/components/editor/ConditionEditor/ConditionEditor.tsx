import { IconInfinity, IconMathFunction } from '@tabler/icons-react';
import { FC } from 'react';

import { Condition } from '~/components/editor/Condition';
import { DEFAULT_CONDITION_DESCRIPTOR } from '~/constants/condition';
import { ConditionAndChain, ConditionGroup } from '~/types/condition';

export interface ConditionEditorProps {
  className?: string;
  conditions?: ConditionGroup;
  onChange?: (newConditions: ConditionGroup) => void;
}

const ConditionEditor: FC<ConditionEditorProps> = (props) => {
  const { className, conditions = [], onChange } = props;

  // Setup
  const hasCondition = !!conditions?.length;

  // Handlers
  const onAddNewVariation = () => {
    onChange?.([...conditions, [DEFAULT_CONDITION_DESCRIPTOR]]);
  };

  const onChainChange = (index: number, newChain: ConditionAndChain) => {
    const newGroup = [...conditions];
    newGroup[index] = newChain;

    onChange?.(newGroup);
  };

  // Markdown
  const renderTitle = () => {
    const Icon = hasCondition ? IconMathFunction : IconInfinity;
    const copy = hasCondition
      ? 'This variation renders under following conditions:'
      : 'This variation renders unconditionally!';

    return (
      <div className="u-py-2x u-font-bold u-text-primary u-tracking-wide">
        <Icon className="u-inline" />
        {copy}
      </div>
    );
  };

  const renderChain = (chain: ConditionAndChain, index: number) => {
    return (
      <Condition
        andChain={chain}
        index={index}
        key={index}
        onChainChange={onChainChange}
      />
    );
  };

  const renderConditions = () => {
    if (!conditions?.length) return null;

    return <div>{conditions.map(renderChain)}</div>;
  };

  return (
    <div className={className}>
      {renderTitle()}
      {renderConditions()}
      <button
        className="u-btn u-btn-outline u-btn-secondary u-uppercase"
        onClick={onAddNewVariation}
        type="button"
      >
        Add new condition
      </button>
    </div>
  );
};

export { ConditionEditor };
