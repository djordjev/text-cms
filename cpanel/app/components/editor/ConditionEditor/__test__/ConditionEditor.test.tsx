import { createRemixStub } from '@remix-run/testing';
import { fireEvent, render, screen } from '@testing-library/react';
import { afterEach, beforeEach } from 'vitest';

import { DEFAULT_CONDITION_DESCRIPTOR } from '~/constants/condition';

import { ConditionEditor, ConditionEditorProps } from '../ConditionEditor';

describe('ConditionEditor', () => {
  let props: ConditionEditorProps;

  const createComponent = () => {
    const RemixStub = createRemixStub([
      {
        path: '/',
        Component: () => <ConditionEditor {...props} />
      }
    ]);

    return <RemixStub />;
  };

  beforeEach(() => {
    props = {
      defaultCondition: [
        [
          ['a1', '=', 'on'],
          ['a2', '=', 'off'],
          ['a3', '>', '3']
        ],
        [['b1', '<', '100']]
      ]
    };
  });

  afterEach(() => {
    vi.resetAllMocks();
  });

  it('renders with conditions', () => {
    const { container } = render(createComponent());

    expect(
      screen.getByText('This variation renders under following conditions:')
    ).toBeInTheDocument();

    const variables = screen.queryAllByRole('textbox', { name: 'Variable' });
    expect(variables).toHaveLength(4);

    fireEvent.click(screen.getByRole('button', { name: 'Add new condition' }));

    const values = screen.queryAllByRole('textbox', { name: 'Value' });
    expect(values).toHaveLength(5);

    const operators = screen.queryAllByRole('combobox', { name: 'operator' });
    expect(operators).toHaveLength(5);

    const hidden = screen.getByTestId('condition') as HTMLInputElement;
    expect(hidden.value).toBe(
      JSON.stringify([
        ...props.defaultCondition!,
        [DEFAULT_CONDITION_DESCRIPTOR]
      ])
    );

    expect(container.childNodes).toMatchSnapshot();
  });

  it('renders empty', () => {
    props.defaultCondition = [];
    const { container } = render(createComponent());

    expect(
      screen.getByText('This variation renders unconditionally!')
    ).toBeInTheDocument();

    expect(
      screen.queryByRole('textbox', { name: 'Value' })
    ).not.toBeInTheDocument();

    fireEvent.click(screen.getByRole('button', { name: 'Add new condition' }));

    const hidden = screen.getByTestId('condition') as HTMLInputElement;
    expect(hidden.value).toBe(JSON.stringify([[DEFAULT_CONDITION_DESCRIPTOR]]));

    expect(container.childNodes).toMatchSnapshot();
  });
});
