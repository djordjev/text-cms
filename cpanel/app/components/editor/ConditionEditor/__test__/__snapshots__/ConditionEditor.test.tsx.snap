// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConditionEditor > renders empty 1`] = `
NodeList [
  <div>
    <div
      class="u-py-2x u-font-bold u-text-primary u-tracking-wide"
    >
      <svg
        class="u-inline"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9.828 9.172a4 4 0 1 0 0 5.656a10 10 0 0 0 2.172 -2.828a10 10 0 0 1 2.172 -2.828a4 4 0 1 1 0 5.656a10 10 0 0 1 -2.172 -2.828a10 10 0 0 0 -2.172 -2.828"
        />
      </svg>
      This variation renders unconditionally!
    </div>
    <button
      class="u-btn u-btn-outline u-btn-secondary u-uppercase"
      type="button"
    >
      Add new condition
    </button>
  </div>,
]
`;

exports[`ConditionEditor > renders with conditions 1`] = `
NodeList [
  <div>
    <div
      class="u-py-2x u-font-bold u-text-primary u-tracking-wide"
    >
      <svg
        class="u-inline"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M3 19a2 2 0 0 0 2 2c2 0 2 -4 3 -9s1 -9 3 -9a2 2 0 0 1 2 2"
        />
        <path
          d="M5 12h6"
        />
        <path
          d="M15 12l6 6"
        />
        <path
          d="M15 18l6 -6"
        />
      </svg>
      This variation renders under following conditions:
    </div>
    <div>
      <div
        class="u-p-2x u-border u-border-solid u-border-info u-rounded-lg u-mb-3xs"
      >
        <div>
          <div
            class="u-flex u-gap-4x u-mb-1x last:u-mb-0"
          >
            <label
              class="u-input u-input-bordered u-flex u-items-center u-gap-2 u-flex-1"
            >
              <svg
                aria-label="Variable"
                class="u-mr-2x"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5 4c-2.5 5 -2.5 10 0 16m14 -16c2.5 5 2.5 10 0 16m-10 -11h1c1 0 1 1 2.016 3.527c.984 2.473 .984 3.473 1.984 3.473h1"
                />
                <path
                  d="M8 16c1.5 0 3 -2 4 -3.5s2.5 -3.5 4 -3.5"
                />
              </svg>
              <input
                class="u-grow"
                placeholder="Enter variable name from payload"
                type="text"
                value="a1"
              />
            </label>
            <select
              aria-label="operator"
              class="u-select u-select-bordered"
            >
              <option>
                =
              </option>
              <option>
                &gt;
              </option>
              <option>
                &gt;=
              </option>
              <option>
                &lt;
              </option>
              <option>
                &lt;=
              </option>
            </select>
            <label
              class="u-input u-input-bordered u-flex u-items-center u-gap-2 u-flex-1"
            >
              <svg
                aria-label="Value"
                class="u-mr-2x"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 9l3 3l-3 3"
                />
                <path
                  d="M3 3m0 2a2 2 0 0 1 2 -2h14a2 2 0 0 1 2 2v14a2 2 0 0 1 -2 2h-14a2 2 0 0 1 -2 -2z"
                />
              </svg>
              <input
                class="u-grow"
                placeholder="Enter required value"
                type="text"
                value="on"
              />
            </label>
            <button
              aria-label="delete-line"
              type="button"
            >
              <svg
                class="tabler-icon tabler-icon-trash"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4 7l16 0"
                />
                <path
                  d="M10 11l0 6"
                />
                <path
                  d="M14 11l0 6"
                />
                <path
                  d="M5 7l1 12a2 2 0 0 0 2 2h8a2 2 0 0 0 2 -2l1 -12"
                />
                <path
                  d="M9 7v-3a1 1 0 0 1 1 -1h4a1 1 0 0 1 1 1v3"
                />
              </svg>
            </button>
          </div>
          <div
            class="u-flex u-gap-4x u-mb-1x last:u-mb-0"
          >
            <label
              class="u-input u-input-bordered u-flex u-items-center u-gap-2 u-flex-1"
            >
              <svg
                aria-label="Variable"
                class="u-mr-2x"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5 4c-2.5 5 -2.5 10 0 16m14 -16c2.5 5 2.5 10 0 16m-10 -11h1c1 0 1 1 2.016 3.527c.984 2.473 .984 3.473 1.984 3.473h1"
                />
                <path
                  d="M8 16c1.5 0 3 -2 4 -3.5s2.5 -3.5 4 -3.5"
                />
              </svg>
              <input
                class="u-grow"
                placeholder="Enter variable name from payload"
                type="text"
                value="a2"
              />
            </label>
            <select
              aria-label="operator"
              class="u-select u-select-bordered"
            >
              <option>
                =
              </option>
              <option>
                &gt;
              </option>
              <option>
                &gt;=
              </option>
              <option>
                &lt;
              </option>
              <option>
                &lt;=
              </option>
            </select>
            <label
              class="u-input u-input-bordered u-flex u-items-center u-gap-2 u-flex-1"
            >
              <svg
                aria-label="Value"
                class="u-mr-2x"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 9l3 3l-3 3"
                />
                <path
                  d="M3 3m0 2a2 2 0 0 1 2 -2h14a2 2 0 0 1 2 2v14a2 2 0 0 1 -2 2h-14a2 2 0 0 1 -2 -2z"
                />
              </svg>
              <input
                class="u-grow"
                placeholder="Enter required value"
                type="text"
                value="off"
              />
            </label>
            <button
              aria-label="delete-line"
              type="button"
            >
              <svg
                class="tabler-icon tabler-icon-trash"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4 7l16 0"
                />
                <path
                  d="M10 11l0 6"
                />
                <path
                  d="M14 11l0 6"
                />
                <path
                  d="M5 7l1 12a2 2 0 0 0 2 2h8a2 2 0 0 0 2 -2l1 -12"
                />
                <path
                  d="M9 7v-3a1 1 0 0 1 1 -1h4a1 1 0 0 1 1 1v3"
                />
              </svg>
            </button>
          </div>
          <div
            class="u-flex u-gap-4x u-mb-1x last:u-mb-0"
          >
            <label
              class="u-input u-input-bordered u-flex u-items-center u-gap-2 u-flex-1"
            >
              <svg
                aria-label="Variable"
                class="u-mr-2x"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5 4c-2.5 5 -2.5 10 0 16m14 -16c2.5 5 2.5 10 0 16m-10 -11h1c1 0 1 1 2.016 3.527c.984 2.473 .984 3.473 1.984 3.473h1"
                />
                <path
                  d="M8 16c1.5 0 3 -2 4 -3.5s2.5 -3.5 4 -3.5"
                />
              </svg>
              <input
                class="u-grow"
                placeholder="Enter variable name from payload"
                type="text"
                value="a3"
              />
            </label>
            <select
              aria-label="operator"
              class="u-select u-select-bordered"
            >
              <option>
                =
              </option>
              <option>
                &gt;
              </option>
              <option>
                &gt;=
              </option>
              <option>
                &lt;
              </option>
              <option>
                &lt;=
              </option>
            </select>
            <label
              class="u-input u-input-bordered u-flex u-items-center u-gap-2 u-flex-1"
            >
              <svg
                aria-label="Value"
                class="u-mr-2x"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 9l3 3l-3 3"
                />
                <path
                  d="M3 3m0 2a2 2 0 0 1 2 -2h14a2 2 0 0 1 2 2v14a2 2 0 0 1 -2 2h-14a2 2 0 0 1 -2 -2z"
                />
              </svg>
              <input
                class="u-grow"
                placeholder="Enter required value"
                type="text"
                value="3"
              />
            </label>
            <button
              aria-label="delete-line"
              type="button"
            >
              <svg
                class="tabler-icon tabler-icon-trash"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4 7l16 0"
                />
                <path
                  d="M10 11l0 6"
                />
                <path
                  d="M14 11l0 6"
                />
                <path
                  d="M5 7l1 12a2 2 0 0 0 2 2h8a2 2 0 0 0 2 -2l1 -12"
                />
                <path
                  d="M9 7v-3a1 1 0 0 1 1 -1h4a1 1 0 0 1 1 1v3"
                />
              </svg>
            </button>
          </div>
          <button
            aria-label="add and condition"
            class="u-btm-sm u-btn-link u-mt-2xs"
            type="button"
          >
            <svg
              class="u-inline u-mr-1xs"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5 4c-2.5 5 -2.5 10 0 16m14 -16c1.38 2.76 2 5.52 1.855 8.448m-11.855 -3.448h1c1 0 1 1 2.016 3.527c.785 1.972 .944 3.008 1.483 3.346"
              />
              <path
                d="M8 16c1.5 0 3 -2 4 -3.5s2.5 -3.5 4 -3.5"
              />
              <path
                d="M16 19h6"
              />
              <path
                d="M19 16v6"
              />
            </svg>
             Add AND condition
          </button>
        </div>
      </div>
      <div
        class="u-p-2x u-border u-border-solid u-border-info u-rounded-lg u-mb-3xs"
      >
        <div>
          <div
            class="u-flex u-gap-4x u-mb-1x last:u-mb-0"
          >
            <label
              class="u-input u-input-bordered u-flex u-items-center u-gap-2 u-flex-1"
            >
              <svg
                aria-label="Variable"
                class="u-mr-2x"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5 4c-2.5 5 -2.5 10 0 16m14 -16c2.5 5 2.5 10 0 16m-10 -11h1c1 0 1 1 2.016 3.527c.984 2.473 .984 3.473 1.984 3.473h1"
                />
                <path
                  d="M8 16c1.5 0 3 -2 4 -3.5s2.5 -3.5 4 -3.5"
                />
              </svg>
              <input
                class="u-grow"
                placeholder="Enter variable name from payload"
                type="text"
                value="b1"
              />
            </label>
            <select
              aria-label="operator"
              class="u-select u-select-bordered"
            >
              <option>
                =
              </option>
              <option>
                &gt;
              </option>
              <option>
                &gt;=
              </option>
              <option>
                &lt;
              </option>
              <option>
                &lt;=
              </option>
            </select>
            <label
              class="u-input u-input-bordered u-flex u-items-center u-gap-2 u-flex-1"
            >
              <svg
                aria-label="Value"
                class="u-mr-2x"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 9l3 3l-3 3"
                />
                <path
                  d="M3 3m0 2a2 2 0 0 1 2 -2h14a2 2 0 0 1 2 2v14a2 2 0 0 1 -2 2h-14a2 2 0 0 1 -2 -2z"
                />
              </svg>
              <input
                class="u-grow"
                placeholder="Enter required value"
                type="text"
                value="100"
              />
            </label>
            <button
              aria-label="delete-line"
              type="button"
            >
              <svg
                class="tabler-icon tabler-icon-trash"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4 7l16 0"
                />
                <path
                  d="M10 11l0 6"
                />
                <path
                  d="M14 11l0 6"
                />
                <path
                  d="M5 7l1 12a2 2 0 0 0 2 2h8a2 2 0 0 0 2 -2l1 -12"
                />
                <path
                  d="M9 7v-3a1 1 0 0 1 1 -1h4a1 1 0 0 1 1 1v3"
                />
              </svg>
            </button>
          </div>
          <button
            aria-label="add and condition"
            class="u-btm-sm u-btn-link u-mt-2xs"
            type="button"
          >
            <svg
              class="u-inline u-mr-1xs"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5 4c-2.5 5 -2.5 10 0 16m14 -16c1.38 2.76 2 5.52 1.855 8.448m-11.855 -3.448h1c1 0 1 1 2.016 3.527c.785 1.972 .944 3.008 1.483 3.346"
              />
              <path
                d="M8 16c1.5 0 3 -2 4 -3.5s2.5 -3.5 4 -3.5"
              />
              <path
                d="M16 19h6"
              />
              <path
                d="M19 16v6"
              />
            </svg>
             Add AND condition
          </button>
        </div>
      </div>
    </div>
    <button
      class="u-btn u-btn-outline u-btn-secondary u-uppercase"
      type="button"
    >
      Add new condition
    </button>
  </div>,
]
`;
